annealing_helper_functions.py

```diff
@@ -76,3 +77,55 @@ def simulated_annealing(graph, function, initial_X, initial_temp, cool, reanneal
                 T = 1.
             
     return X, history
+
+def prob(a, b):
+    return np.exp(a / b)
+
+# Adapted from Lecture14_Parallel_Tempering_and_Emcee.ipynb
+def parallel_tempering(graph, function, initial_Xs, initial_temps, 
+                       iterr, swap_function, nswaps, nbefore):
+    # Make sure inputs are ok
+    assert(len(initial_temps) == len(initial_Xs)), "Mismatched input dimensions"
+    assert(initial_temps[0] == 1), "First temperature should be one"
+
+    # Initialize stuff
+    nsystems = len(initial_temps)
+    Xs = initial_Xs
+    Ts = initial_temps
+    P = [lambda d: np.exp(d / Ts[c]) for c in range(nsystems)]
+    prev_Es = [function(graph, Xs[i]) for i in range(nsystems)]
+    history = [[] for i in xrange(nsystems)]
+
+    for c in xrange(nsystems):
+        history[c].append(prev_Es[c])
+
+    for step in range(iterr):
+        # Run nbefore steps of simulated annealing
+        X_stars = [swap_function(Xs[i], nswaps) for i in xrange(nsystems)]
+        new_Es = [function(graph, X_stars[i]) for i in xrange(nsystems)]
+        delta_Es = [i - j for i, j in zip(new_Es, prev_Es)]
```

Comment:
Agreed on both fronts. I used function-in-list following Pavlos' implementation, but yeah, it would be better to have a function P that takes temperature and delta E as an argument.